brickstrap-components.sh

```diff
@@ -0,0 +1,215 @@
+    # permit a blanket override of 'true' to accept any path
+    true) return 0;;
+    # permit a blanket override of 'false' to disallow any path
+    false) return 1;;
+    *)
+        # premit arbitrary path tests through a callback interface
+        # this can be used to reject files based on e.g. a static blacklist
+        if [ -n "$2" ]; then
+            eval "$2" "$1" || return 1
+        else
+            return 1
+        fi
+    ;;
+    esac
+}
+
+#
+# Checks a path against a path test, and if the test succeeds outputs the path.
+#
+# $1 component to test
+# $2 base path to test
+# $3 path test to perform
+#
+function brp_check_path()
+{
+    if [ $# -ge 3 ] && brp_accept_path_if "$(br_project_dir)/$1/$2" "$3"; then
+        echo "$(readlink -f "$(br_project_dir)/$1/$2")"
+        return 255
+    else
+        return 0
+    fi
+}
+
+#
+# Functions to generate a list of paths matching a given base path and path
+# test.
+#
+# The paths may be consumed by piping output into a 'while IFS='' read -r line'
+# block or by using the callback interface of br_consume_path_list* functions.
+#
+# $1 base path to find
+# $2 optional: path test. If not specified, a default test of '-e' is assumed.
+#
+function br_list_paths()
+{
+    if [ $# -eq 0 -o -z "$1" ]; then
+        return 1
+    fi
+    if [ $# -eq 1 ]; then
+        brp_iterate_components brp_check_path "$1" -e
+    elif [ -n "$2" ]; then
+        brp_iterate_components brp_check_path "$@"
+    else
+        return 1
+    fi
+    if [ $? -eq 255 ]; then
+        return 0
+    else
+        return 1
+    fi
+}
+
+#
+# Convenience version of br_list_paths which uses a fixed path test of '-d'.
+#
+# $1 base path to find
+#
+function br_list_directories()
+{
+    [ $# -ge 1 ] && br_list_paths "$1" -d
+}
+
+#
+# Callback interface to consume path lists generated by br_list_*(). Usage:
+#
+# br_list_* path ...optional_list_args | \
+#     br_consume_path_list* call_back_name optional_call_back_args...
+#
+# The calling convention is:
+#
+# The callback argument is eval'ed; the callback is passed the current file
+# being consumed (BR_PATHS_CUR_FILE) as last argument. Optional callback args
+# are passed on.
+#
+# No special handling for spaces-in-arguments is performed, so users of
+# br_consume_path_list* functions must make sure any spaces in arguments are
+# properply escaped.
+#
+# Return code of the callback is captured (if non-zero). If previous
+# invocations of the callback failed, the last captured error code will be
+# saved in BR_PATHS_CB_RETURNCODE. If any invocations of the callback fail
+# (i.e. returned non-zero status code) or if another error occurs, the
+# br_consume_path_list*() function will return with a non-zero return code.
+#
+# This means something like this will print directories found using a given
+# 'printf' format:
+#
+# br_list_directories query | \
+#    br_consume_path_list printf 'found: %s\\n' || echo "error"
+#
+
```

Comment:
The purpose of the general mechanism is to execute a callback for each path found but be able to fail if any one of the callback invocations fail or if the mapping mechanism itself fails.

The thing being consumed is the produced list (line-by-line) of matching paths. If you like I could rename `consume` into `map`. Essentially the idea is: produce | map callback

For plain files you can see the `cat` version to cat them together (useful for e.g. `config` or `tar-exclude`). For directories you can see the `_iterate_directories` version (useful for e.g. running all hooks from all matching hooks directories, or examining all package files from all packages directories).